beacon-chain/slasher/receive_test.go

```diff
@@ -5,11 +5,129 @@ import (
 	"testing"
 
 	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	dbtest "github.com/prysmaticlabs/prysm/beacon-chain/db/testing"
 	slashertypes "github.com/prysmaticlabs/prysm/beacon-chain/slasher/types"
 	"github.com/prysmaticlabs/prysm/shared/event"
 	"github.com/prysmaticlabs/prysm/shared/testutil/require"
+	logTest "github.com/sirupsen/logrus/hooks/test"
 )
 
+func Test_processQueuedAttestations_DetectsSurroundingVotes(t *testing.T) {
+	hook := logTest.NewGlobal()
+	beaconDB := dbtest.SetupDB(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &Service{
+		serviceCfg: &ServiceConfig{
+			Database: beaconDB,
+		},
+		params:           DefaultParams(),
+		attestationQueue: make([]*slashertypes.CompactAttestation, 0),
+	}
+	currentEpochChan := make(chan uint64, 0)
+	exitChan := make(chan struct{})
+	go func() {
+		s.processQueuedAttestations(ctx, currentEpochChan)
+		exitChan <- struct{}{}
+	}()
+	s.attestationQueue = []*slashertypes.CompactAttestation{
+		{
+			AttestingIndices: []uint64{0, 1},
+			Source:           1,
+			Target:           2,
+			SigningRoot:      [32]byte{1},
+		},
+		{
+			AttestingIndices: []uint64{0, 1},
+			Source:           0,
+			Target:           3,
+			SigningRoot:      [32]byte{1},
+		},
+	}
+	currentEpoch := uint64(4)
+	currentEpochChan <- currentEpoch
+	cancel()
+	<-exitChan
+	require.LogsContain(t, hook, "Slashable offenses found")
+	require.LogsContain(t, hook, "Attester surrounding vote")
+}
+
+func Test_processQueuedAttestations_DetectsSurroundedVote(t *testing.T) {
+	hook := logTest.NewGlobal()
+	beaconDB := dbtest.SetupDB(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &Service{
+		serviceCfg: &ServiceConfig{
+			Database: beaconDB,
+		},
+		params:           DefaultParams(),
+		attestationQueue: make([]*slashertypes.CompactAttestation, 0),
+	}
+	currentEpochChan := make(chan uint64, 0)
+	exitChan := make(chan struct{})
+	go func() {
+		s.processQueuedAttestations(ctx, currentEpochChan)
+		exitChan <- struct{}{}
+	}()
+	s.attestationQueue = []*slashertypes.CompactAttestation{
+		{
+			AttestingIndices: []uint64{0, 1},
+			Source:           0,
+			Target:           3,
+			SigningRoot:      [32]byte{1},
+		},
+		{
+			AttestingIndices: []uint64{0, 1},
+			Source:           1,
+			Target:           2,
+			SigningRoot:      [32]byte{1},
+		},
+	}
+	currentEpoch := uint64(4)
+	currentEpochChan <- currentEpoch
+	cancel()
+	<-exitChan
+	require.LogsContain(t, hook, "Slashable offenses found")
+	require.LogsContain(t, hook, "Attester surrounded vote")
+}
+
+func Test_processQueuedAttestations_NotSlashable(t *testing.T) {
```

Comment:
Would it make sense to test more not slashable scenarios (e.g. (0,2), (1,3)) or is this already unit tested somewhere else?

Reply:
Yes indeed we can add more